### PR TITLE
Feature/xmldoc stratis.bitcoin.utilities

### PR DESCRIPTION
**Pull request by Aprogiena**
XML documentation comments for Stratis.Bitcoin.Utilities files. Also fixes a code style in many places according to the current guideline. And it also includes a few TODOs to mark the code that should be reviewed later - this includes potentially dangerous code, or a code that is not used and could be removed from the code base.

All code changes by this PR should be equivalent to the original code - i.e. the behavior of all the code should be unchanged by this PR.

### CODE REVIEW

File: Stratis.Bitcoin/Utilities/Extensions.cs [hunk 1]
```diff
@@ -3,26 +3,69 @@
...
 namespace Stratis.Bitcoin.Utilities
 {
-	public static class TimeSpans
+    /// <summary>
+    /// Commonly used time spans.
+    /// </summary>
+    public static class TimeSpans
     {
+        /// <summary>Time span of 100 milliseconds.</summary>
+        /// <remarks>TODO: Usually milliseconds are shorted to "ms", not "mls", so I suggest renaming this to "Ms100".</remarks>
```
Comment (dangershony) [inline Stratis.Bitcoin/Utilities/Extensions.cs]: yeap

File: Stratis.Bitcoin/Utilities/PeriodicTask.cs [hunk 2]
```diff
@@ -4,50 +4,75 @@
...
+        /// <param name="cancellation">Cancellation token that allows the caller to stop the loop.</param>
+        /// <param name="refreshRate">Delay between two executions of the task.</param>
+        /// <param name="delayStart">If true, the first execution of the task is made after the <paramref name="refreshRate"/> time.</param>
+        /// <returns>This task to enable fluent code.</returns>
         PeriodicTask Start(CancellationToken cancellation, TimeSpan refreshRate, bool delayStart = false);
     }
 
+    /// <inheritdoc />
+    /// <remarks>This implementation spawns a new thread for each new periodic task.</remarks>
     public class PeriodicTask : IPeriodicTask
```
Comment (dangershony) [inline Stratis.Bitcoin/Utilities/PeriodicTask.cs]: The plan is to get rid of this class and only use the AsyncLoop, no point in holding a task that sleeps 90% of the time.

File: Stratis.Bitcoin/Utilities/AsyncLock.cs [hunk 3]
```diff
@@ -9,81 +9,136 @@
...
-		public CancellationTokenSource Cancellation { get; private set; }
-		private readonly ConcurrentExclusiveSchedulerPair schedulerPair; // reference kept for perf counter
+        /// <inheritdoc />
+        /// <remarks>TODO: Why is this public? No one outside seems to be using this.</remarks>
+        public CancellationTokenSource Cancellation { get; private set; }
+
+        /// <summary>Provider of exclusive and concurrent scheduler.</summary>
+        /// <remarks>
+        /// Reference is kept for performance counter. 
+        /// <para>TODO: It seems that this is not used anywhere. What performance counter?</para>
```
Comment (Aprogiena) [inline Stratis.Bitcoin/Utilities/AsyncLock.cs]: so should i remove TODO then? 
or should i remove the field altogether?

File: Stratis.Bitcoin/Utilities/AsyncLock.cs [hunk 4]
```diff
@@ -9,81 +9,136 @@
...
-    /// An async reader writer lock for concurrent and exclusive work
+    /// An async reader writer lock for concurrent and exclusive work.
+    /// <para>
+    /// The class uses ConcurrentExclusiveSchedulerPair to access two task schedulers - concurrent 
+    /// scheduler and exclusive scheduler. The exclusive scheduler guarantees only one task to be run 
+    /// at the same, which is what is used as a writer lock. The concurrent scheduler allows multiple 
+    /// tasks to run simultaneously, but the exclusivity of exclusive scheduler is respected, so it is 
+    /// used as a reader lock.
+    /// </para>
+    /// <para>
```
Comment (Aprogiena) [inline Stratis.Bitcoin/Utilities/AsyncLock.cs]: yes, i just need to move this warning to remarks section

File: Stratis.Bitcoin/Utilities/AsyncLoopFactory.cs [hunk 5]
```diff
@@ -5,35 +5,92 @@
 
 namespace Stratis.Bitcoin.Utilities
 {
+    /// <summary>
+    /// Factory for creating and also possibly starting application defined tasks inside async loop.
+    /// </summary>
```
Comment (Aprogiena) [inline Stratis.Bitcoin/Utilities/AsyncLoopFactory.cs]: no problem with that, but probably different task/issue

File: Stratis.Bitcoin/Utilities/AsyncLock.cs [hunk 6]
```diff
@@ -9,81 +9,136 @@
...
-		/// Queue sequential work to the ExclusiveScheduler.
-		/// Delegates calling this method will be done in sequentially, 
-		/// the first task will be queued on the Default scheduler subsequent exclusive tasks will run in that same thread.
-		/// </summary>
-		public Task<T> WriteAsync<T>(Func<T> func)
-		{
-			return this.exclusiveFactory.StartNew(func, this.Cancellation.Token);
-		}
+        /// <inheritdoc />
+        /// <remarks>See warning in <see cref="AsyncLock"/> remarks section.</remarks>
```
Comment (Aprogiena) [inline Stratis.Bitcoin/Utilities/AsyncLock.cs]: That seems to be correct, removing.

File: Stratis.Bitcoin/Utilities/CustomThreadPoolScheduler.cs [hunk 7]
```diff
@@ -8,131 +8,196 @@
...
+        /// the task there are no more tasks to execute or being executed at the moment. However, assume a case in which no one 
+        /// waits on this event. This means that after first task completes, no one resets the event and even if there later are 
+        /// many tasks being executed or in the queue, this event would still be set until someone would consume it, which would 
+        /// immediately cause the consumer to think that there is no more work to be done.
+        /// <para>
+        /// Moreover, there seem to be no synchronization over the shared resource among multiple threads, 
+        /// so race conditions are possible.
+        /// </para>
+        /// <para>
+        /// As no one is using <see cref="WaitFinished"/> anyway, I suggest to remove both, this event and the method using it
```
Comment (Aprogiena) [inline Stratis.Bitcoin/Utilities/CustomThreadPoolScheduler.cs]: Issue created for this

File: Stratis.Bitcoin/Utilities/CustomThreadPoolScheduler.cs [hunk 8]
```diff
@@ -8,131 +8,196 @@
...
+            {
+                return (this.threadCount - this.AvailableThreads) + this.QueuedCount;
+            }
+        }
+
+        /// <summary>
+        /// Thread procedure that consumes the tasks added to the queue and executes them.
+        /// <para>There are <see cref="threadCount"/> number of threads executing this method at the same time.</para>
+        /// </summary>
+        void Do()
```
Comment (Aprogiena) [inline Stratis.Bitcoin/Utilities/CustomThreadPoolScheduler.cs]: It is, but tend to only rename stuff while xmldocing when it is against code style.
Do you want me to change it here for this PR?

Comment (Aprogiena) [inline Stratis.Bitcoin/Utilities/CustomThreadPoolScheduler.cs]: I see the same methods name elsewhere in this PR (DBreezeSingleThreadSession).

File: Stratis.Bitcoin/Utilities/DBreezeSingleThreadSession.cs [hunk 9]
```diff
@@ -18,60 +18,105 @@ public interface IDBreezeSingleThreadSession : IDisposable
 
     public class DBreezeSingleThreadSession : IDBreezeSingleThreadSession
     {
-		public DBreezeSingleThreadSession(string threadName, string folder)
+        /// <summary>Access to DBreeze database.</summary>
+        private DBreezeEngine engine;
+
+        /// <summary>
+        /// Scheduler that uses only a single thread to provide ability for exclusive execution of the tasks. 
+        /// This is used as an alternative to lock-based critical section implementation.
```
Comment (Aprogiena) [inline Stratis.Bitcoin/Utilities/DBreezeSingleThreadSession.cs]: tx for this info, rephrasing and adding to remarks section

File: Stratis.Bitcoin/Utilities/Extensions.cs (hunk 1 above)
Comment (Aprogiena) [inline Stratis.Bitcoin/Utilities/Extensions.cs]: as this is not used elsewhere, i will change it as a part of this PR

File: Stratis.Bitcoin/Utilities/PeriodicTask.cs (hunk 2 above)
Comment (Aprogiena) [inline Stratis.Bitcoin/Utilities/PeriodicTask.cs]: ok for another task/issue

**Comment by Aprogiena** [comment]
Spotted couple of forgotten uncommented things, adding ...

File: Stratis.Bitcoin/Utilities/AsyncLock.cs (hunk 3 above)
Comment (dangershony) [inline Stratis.Bitcoin/Utilities/AsyncLock.cs]: Remove the field

File: Stratis.Bitcoin/Utilities/CustomThreadPoolScheduler.cs (hunk 8 above)
Comment (dangershony) [inline Stratis.Bitcoin/Utilities/CustomThreadPoolScheduler.cs]: yeah different PR

File: Stratis.Bitcoin/Utilities/AsyncLock.cs (hunk 3 above)
Comment (Aprogiena) [inline Stratis.Bitcoin/Utilities/AsyncLock.cs]: ok

**Pull request merged**